controller/konnect/konnectextension_controller.go

```diff
@@ -219,7 +227,49 @@ func (r *KonnectExtensionReconciler) Reconcile(ctx context.Context, req ctrl.Req
 		sdkops.SDKToken(token),
 	)
 
-	cp, err := ops.GetControlPlaneByID(ctx, sdk.GetControlPlaneSDK(), *ext.Spec.KonnectControlPlane.ControlPlaneRef.KonnectID)
+	var konnectCPID string
+	switch ext.Spec.KonnectControlPlane.ControlPlaneRef.Type {
+	case commonv1alpha1.ControlPlaneRefKonnectNamespacedRef:
+		cpRef := ext.Spec.KonnectControlPlane.ControlPlaneRef.KonnectNamespacedRef
+		cpNamepace := ext.Namespace
+		// TODO: get namespace from cpRef.Namespace when allowed to reference CP from another namespace.
+		kgcp := &konnectv1alpha1.KonnectGatewayControlPlane{}
+		err := r.Client.Get(ctx, client.ObjectKey{
+			Namespace: cpNamepace,
+			Name:      cpRef.Name,
+		}, kgcp)
+
+		if err != nil {
+			_, errPatch := patch.StatusWithCondition(
+				ctx, r.Client, &ext,
+				consts.ConditionType(konnectv1alpha1.ControlPlaneRefValidConditionType),
+				metav1.ConditionFalse,
+				consts.ConditionReason(konnectv1alpha1.ControlPlaneRefReasonInvalid),
+				err.Error(),
+			)
+			return ctrl.Result{}, errPatch
+		}
+
+		if !lo.ContainsBy(kgcp.Status.Conditions, func(cond metav1.Condition) bool {
+			return cond.Type == konnectv1alpha1.KonnectEntityProgrammedConditionType &&
+				cond.Status == metav1.ConditionTrue
+		}) {
+			_, errPatch := patch.StatusWithCondition(
+				ctx, r.Client, &ext,
+				consts.ConditionType(konnectv1alpha1.ControlPlaneRefValidConditionType),
+				metav1.ConditionFalse,
+				consts.ConditionReason(konnectv1alpha1.ControlPlaneRefReasonInvalid),
+				fmt.Sprintf("Konnect control plane %s/%s not programmed yet", cpNamepace, cpRef.Name),
+			)
+			// Update of KonnectGatewayControlPlane will trigger reconciliation of KonnectExtension.
+			return ctrl.Result{}, errPatch
+		}
+		konnectCPID = kgcp.GetKonnectID()
+	case commonv1alpha1.ControlPlaneRefKonnectID:
+		konnectCPID = *ext.Spec.KonnectControlPlane.ControlPlaneRef.KonnectID
+	}
+
+	cp, err := ops.GetControlPlaneByID(ctx, sdk.GetControlPlaneSDK(), konnectCPID)
```

Comment:
We should get the CP from Konnect only when the CP is referenced through KonnectID. In case the reference is via `NamespacedRef`, having it programmed is enough of a check.
I think we should move this to the `commonv1alpha1.ControlPlaneRefKonnectID` switch case.

Reply:
fetching CP from Konnect is also required for setting `clusterType` and `endpoints` in status of the `KonnectExtesion` because we need the values. Do you think we should also fetch the values from `cp.Status` in such case?

Reply:
Ok, I've opened an issue to address that: https://github.com/Kong/gateway-operator/issues/1275. We can move this one forward in the meantime 👍🏻